google-http-client-apache-v5/src/main/java/com/google/api/client/http/apache/v5/Apache5HttpResponse.java

```diff
@@ -0,0 +1,120 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.api.client.http.apache.v5;
+
+import com.google.api.client.http.LowLevelHttpResponse;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.apache.hc.client5.http.classic.methods.HttpUriRequestBase;
+import org.apache.hc.core5.http.ClassicHttpResponse;
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.core5.http.HttpEntity;
+import org.apache.hc.core5.http.message.StatusLine;
+
+final class Apache5HttpResponse extends LowLevelHttpResponse {
+
+  private static final Logger LOGGER = Logger.getLogger(Apache5HttpResponse.class.getName());
+  private final HttpUriRequestBase request;
+  private final ClassicHttpResponse response;
+  private final Header[] allHeaders;
+  private final HttpEntity entity;
+  private final Apache5ResponseContent content;
+
+  Apache5HttpResponse(HttpUriRequestBase request, ClassicHttpResponse response) {
+    this.request = request;
+    this.response = response;
+    this.allHeaders = response.getHeaders();
+    this.entity = response.getEntity();
+    Apache5ResponseContent contentValue = null;
+    if (entity != null) {
+      try {
+        contentValue = new Apache5ResponseContent(entity.getContent(), response);
+      } catch (IOException ex) {
+        LOGGER.log(Level.SEVERE, "Error when obtaining content from the response", ex);
+      }
+    }
+    this.content = contentValue;
+  }
+
+  @Override
+  public int getStatusCode() {
+    return response.getCode();
+  }
+
+  @Override
+  public InputStream getContent() throws IOException {
+    return content;
+  }
+
+  @Override
+  public String getContentEncoding() {
+    return entity != null ? entity.getContentEncoding() : null;
+  }
+
+  @Override
+  public long getContentLength() {
+    return entity == null ? -1 : entity.getContentLength();
+  }
+
+  @Override
+  public String getContentType() {
+    return entity == null ? null : entity.getContentType();
+  }
+
+  @Override
+  public String getReasonPhrase() {
+    return response.getReasonPhrase();
+  }
+
+  @Override
+  public String getStatusLine() {
+    return new StatusLine(response).toString();
+  }
+
+  public String getHeaderValue(String name) {
+    return response.getLastHeader(name).getValue();
+  }
+
+  @Override
+  public int getHeaderCount() {
+    return allHeaders.length;
+  }
+
+  @Override
+  public String getHeaderName(int index) {
+    return allHeaders[index].getName();
+  }
+
+  @Override
+  public String getHeaderValue(int index) {
+    return allHeaders[index].getValue();
+  }
+
+  /** Aborts execution of the request. */
+  @Override
+  public void disconnect() {
+    request.abort();
+    try {
+      // this call also close the response
+      content.close();
```

Comment:
Do we need to do this? This wasn't part of the Apache 4 contract.

Reply:
True, I was worried about closing the response, but I see it's better to close the response directly to be consistent with the old contract.

Reply:
Note that `ClassicHttpResponse` also [has an endpoint to disconnect](https://github.com/apache/httpcomponents-client/blob/ec4da301bca5b5fd8c34a228cb43738d043e6160/httpclient5/src/main/java/org/apache/hc/client5/http/impl/classic/CloseableHttpResponse.java#L206). I believe it'd be helpful to close this response as well.

Reply:
Discussed offline.